### ISSUE TITLE
Rendered object is not reflective of actual underlying data

### ISSUE THREAD

**Issue by davesargrad** [konvajs/react-konva#730]
I've got an object that has several elements (rectangles, paths, etc). I overlay edit handles when I want to edit the object.

In the following view we see an object with green resize handles and a pink rectangle that acts as a reposition handle (to drag the entire object).
![image](https://user-images.githubusercontent.com/48793689/217240704-9713c927-1bf0-4260-b5c0-b3a413aa5039.png)

I store the object geometry in a backend redux store. I then render the object and the edit handles as follows:
```
return (
    <React.Fragment>
      {itemView}
      {editView}
    </React.Fragment>
  );
```

The itemView and the editView should overlay/align. Whenever I drag a resize handle, or drag the reposition handle, I immediately compute a new position/size for the item, and I dispatch an update to the redux store. I then rerender the set of items on my canvas.

In the following we see a view where the edit elements are out of sync with the item elements. Here I've dragged the pink "resposition handle" quickly and the position of the pink box gets misaligned (not only relative to the itemView, but also relative to other handles in the editView). 

![image](https://user-images.githubusercontent.com/48793689/217238378-962cca43-46fc-4f3d-adb6-86a7c3a05a87.png)

This misalignment happens whether its the green resize handles, or the pink reposition handle that I drag. 
![image](https://user-images.githubusercontent.com/48793689/217250446-40b5b3e9-4992-4d89-b970-a61b207fa81b.png)

If I drag slowly then they dont get as visibly out of sync.

The itemView in the following image is simply a grey rectangle. I've enabled it here to further illustrate the misalignment. 
![image](https://user-images.githubusercontent.com/48793689/217238813-e1f52eb6-de8b-4d05-9db0-d4f17d856780.png)

If I simply refresh by toggling the visibility of the editView (pink reposition handle) then it aligns properly (so prior to that "visibility toggle action" the editView was stale relative to the backing redux store containing the actual object detail)
![image](https://user-images.githubusercontent.com/48793689/217238974-74db6c99-f599-4d53-a6fb-33eb95bf7d7e.png)

So its the editView (pink reposition handle, and green resize handles) that seem to not be re-rendered properly when the backing redux store is updated. Its state seems to be inconsistent (stale) with respect to the backing store. Whereas the itemView state is consistent.

I've knocked my head against this for a couple days now. **Any guidance would be appreciated.**

-------------------------------------
Some additional code detail follows (provided to assist you in understanding my current design).

The editView is currently formed as follows:
```
let editView = (
    <>
      {
        lockDrag?<></>:
          <Group 
            x={iBnds.x}
            y={iBnds.y}
          >
            <FrameItemDragHandle 
                  x={0}
                  y={0}
                  width={iBnds.width}
                  height={iBnds.height}
                  frameWidth={fw}
                  frameHeight={fh}
                  lock={lockDrag}
                  itemBounds={iBnds}
                  onDragStart={() => onDragStart(frameId, itemId)}
                  onDragMove={(nib) => onDragMove(frameId, itemId, itemUC(nib.x / fw, nib.y / fh, nib.width / fw, nib.height / fh))}
                  onDragEnd={(nib) => onDragEnd(frameId, itemId, itemUC(nib.x / fw, nib.y / fh, nib.width / fw, nib.height / fh))}
                  style={itemStyle.boundary} />
          </Group>
      }
      {
        lockResize?<></>:
        <Group 
          x={iBnds.x}
          y={iBnds.y}
        >
          {handles.map((hdl, hidx) => {
            return (
              <FrameItemResizeHandle 
                key={hidx} 
                handle={hdl}
                frameWidth={fw}
                frameHeight={fh}
                itemBounds={iBnds}
                onDragStart={() => onResizeStart(frameId, itemId)}
                onDragMove={(nib) => onResizeMove(frameId, itemId, itemUC(nib.x / fw, nib.y / fh, nib.width / fw, nib.height / fh))}
                onDragEnd={(nib) => onResizeEnd(frameId, itemId, itemUC(nib.x / fw, nib.y / fh, nib.width / fw, nib.height / fh))}
                style={itemStyle.resizeHandle} />
              )
          })}
        </Group>
      }
    </>
  )
```

The itemView currently is formed as follows:
```
let itemView = (
    <>
      {
        !lockDrag && !lockResize?<></>:
        <Group 
          draggable={false}
          onMouseEnter={(e) => onMouseEnter(e, frameId, itemId, item)}
          onMouseLeave={(e) => onMouseLeave(e, frameId, itemId, item)}
          onDragStart={() => onDragStart(frameId, itemId)}
          onDragMove={(e) => onDragMove(frameId, itemId, itemUC(e.target.x()/fw, e.target.y()/fh, item.width, item.height))}
          onDragEnd={(e) => onDragEnd(frameId, itemId, itemUC(e.target.x()/fw, e.target.y()/fh, item.width, item.height))}
          x={iBnds.x}
          y={iBnds.y}
        >
          <Path
            closed={true}
            draggable={false}
            x={0}
            y={0}
            points={pts}
            style={itemStyle.boundingBox}
          />
          {true?<></>:<Path
            closed={true}
            draggable={false}
            x={0}
            y={0}
            points={typeof item.boundary !== 'undefined'?item.boundary.map((ub, ubi) => ((ubi % 2) === 0) ? ub * fw: ub * fh):[]}
            style={itemStyle.boundary}
          />}
        </Group>
      }
    </>
  )
```
The current resize handle design:

```
const boundPosition = (hdl, pos, iBnds) => {
  if(hdl.name === "top") {
    return {x: iBnds.x+hdl.x, y: Math.min(pos.y, iBnds.y+hdl.y+iBnds.height)}
  } else if(hdl.name === "bottom") {
    return {x: iBnds.x+hdl.x, y: Math.max(pos.y, iBnds.y+hdl.y-iBnds.height)}
  } else if(hdl.name === "left") {
    return {x: Math.min(pos.x, iBnds.x+hdl.x+iBnds.width), y: iBnds.y+hdl.y}
  } else if(hdl.name === "right") {
    return {x: Math.max(pos.x, iBnds.x+hdl.x-iBnds.width), y: iBnds.y+hdl.y}
  }
  return pos;
}

const computeNewItemBounds = (hdlX, hdlY, hdl, iBnds) => {
  var bnds = {x: 100, y: 100};
  if(hdl.name === "top") {
    const ny = iBnds.y + hdlY + hdl.sz/2;
    const nh = iBnds.y - ny + iBnds.height;
    bnds = {x: iBnds.x, y: ny, width: iBnds.width, height: nh}
  } else if(hdl.name === "bottom") {
    const ny = iBnds.y;
    const nh = hdlY + hdl.sz/2;
    bnds = {x: iBnds.x, y: ny, width: iBnds.width, height: nh}
  } else if(hdl.name === "left") {
    const nx = iBnds.x + hdlX + hdl.sz/2;
    const nw = iBnds.x - nx + iBnds.width;
    bnds = {x: nx, y: iBnds.y, width: nw, height: iBnds.height}
  } else if(hdl.name === "right") {
    const nx = iBnds.x;
    const nw = hdlX + hdl.sz/2;
    bnds = {x: nx, y: iBnds.y, width: nw, height: iBnds.height}
  }
  return bnds;
}

const FrameItemResizeHandle = ({handle, frameWidth, frameHeight, itemBounds, style, dragBoundFunc, onDragStart, onDragMove, onDragEnd}) => {
  const {x, y, sz, name} = handle;
  return (
    <Rectangle
        draggable={true}
        x={x}
        y={y}
        width={sz}
        height={sz}
        style={style}
        onDragStart={(e) => onDragStart(e, name)}
        onDragMove={(e) => onDragMove(computeNewItemBounds(e.target.x(), e.target.y(), handle, itemBounds))}
        onDragEnd={(e) => onDragEnd(computeNewItemBounds(e.target.x(), e.target.y(), handle, itemBounds))}
        dragBoundFunc={(pos) => boundPosition(handle, pos, itemBounds)}
    />
  )
}
```

The current reposition handle design:
```
const computeNewItemBounds = (hdlX, hdlY, iBnds, fw, fh) => {
  var bnds = {x: 200, y: 100};
  bnds = {x: iBnds.x + hdlX, y: iBnds.y + hdlY, width: iBnds.width, height: iBnds.height}

  const u_bnds = {x: (iBnds.x + hdlX)/fw, y: (iBnds.y + hdlY)/fh, width: iBnds.width/fw, height: iBnds.height/fh}
    const newItem = cTL2C(u_bnds);
    if(iVUCI(newItem)) {
      return bnds;
    }
    return iBnds;
}

const FrameItemDragHandle = ({x, y, width, height, frameWidth, frameHeight, lock, itemBounds, style, dragBoundFunc, onDragStart, onDragMove, onDragEnd}) => {

  return (
    <Rectangle
        draggable={!lock}
        x={x}
        y={y}
        width={width}
        height={height}
        style={style}
        onDragStart={(e) => onDragStart(e)}
        onDragMove={(e) => onDragMove(computeNewItemBounds(e.target.x(), e.target.y(), itemBounds, frameWidth, frameHeight))}
        onDragEnd={(e) => onDragEnd(computeNewItemBounds(e.target.x(), e.target.y(), itemBounds, frameWidth, frameHeight))}

    />
  )
}
```
-----------------------------------------------------
Update after continuing to struggle with this:

This is a perfect view .. In this view the FrameItem object is a Group containing several react-konva elements: an Html menu item in upper left corner, a pink drag Rect, and a set of green Resize handles. The drag Rect and Resize handles are themselves in a child Group. 

![image](https://user-images.githubusercontent.com/48793689/218749493-559492df-c98c-4b19-accc-c4e42c427016.png)

The pink Rect was used to drag the entire structure. Using redux, I triggered a state update through the use of onDragMove, and onDragEnd. The object was re-rendered since it does so on update of the backing object state. The object was then re-rendered. The menu control in upper left and the green resize handles are in a state that is consistent with the backing state. However the pink drag box is not in the right position, even though the console log has the appropropriate values for x, y, width, and height

I've also kept track of the revision to the object. 

In the following console view this is seen as the 863 revision to this displayed item.
![image](https://user-images.githubusercontent.com/48793689/218750038-f8d7481b-856d-4f78-9cbd-2b44c66d323c.png)

The x, y, width, and height are in "unified coordinates [0 to 1] relative to the size of the underlying image. The x, y, width, and height passed to the objects rendered within a Layer and Group are computed from these values.

The green resize handles and the bluish menu control (in upper left) are in the proper position relative to x, y, width, and height. However the pink Rect is in a position that is inconsistent with these x, y, width, and height values. 

If I force a refresh (which I can do simply by toggling the draggability of the pink box), then the pink box snaps into the proper position.

![image](https://user-images.githubusercontent.com/48793689/218750886-da722b37-8706-4ce8-b7ad-8243a04faab6.png)

So in short, the pink drag box is being rendered within the react hierarchy, but its depicted state is inconsistent with the X, Y, WIDTH, and HEIGHT values passed to the Konva Rect

**Comment by davesargrad** [comment]
Initially I thought this was just a problem in the way that I was using react-konva. I've now further instrumented with diagnostics. I'm starting to believe this is a rendering/refresh bug.

I'll post more detail in the body of the issue, above.

**Comment by davesargrad** [comment]
I am testing useStrictMode.. This may be my solution

**Comment by davesargrad** [comment]
Closed with the introduction of useStrictMode(true)... I do believe that this warrants an update to the react-konva documentation, to provide a better context for when useStrictMode(true) is absolutely required. Though the existing documentation is already very good, I think this is a case where it can be even better, simply by giving a couple scenarios where useStrictMode(true).. is essential to ensure that the displayed stage is properly synchronized with the applications backend state
